auth_rust/src/main.rs

```diff
@@ -0,0 +1,230 @@
+
+    if counter_limit > token_value.monthly_limit {
+        return (
+            StatusCode::TOO_MANY_REQUESTS,
+            [(header::CONTENT_TYPE, "application/json")],
+            r#"{"message":"Monthly limit exceeded"}"#,
+        );
+    }
+
+    match auth_repo.increment_monthly_counter(&token_value, counter_limit).await {
+        Ok(_) => (),
+        Err(_) => {
+            return (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                [(header::CONTENT_TYPE, "application/json")],
+                r#"{"message":"Internal server error"}"#,
+            );
+        }
+    }
+
+    Metric::count("pesto.successful_auth")
+        .with_tag("user_email", token_value.user_email.clone())
+        .send();
+    (StatusCode::OK, [(header::CONTENT_TYPE, "text/plain")], "")
+}
+
+enum AuthError {
+    TokenNotRegistered,
+    FailedToAcquireToken(RedisError),
+}
+
+impl Display for AuthError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            AuthError::TokenNotRegistered => write!(f, "Token not registered"),
+            AuthError::FailedToAcquireToken(e) => write!(f, "Failed to acquire token: {}", e),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+struct TokenValue {
+    user_email: String,
+    monthly_limit: i64,
+    revoked: bool,
+}
+
+#[derive(Clone)]
+struct AuthRepo<Command: AsyncCommands> {
+    pub redis_client: Command,
+}
+
+impl<Command: AsyncCommands> Debug for AuthRepo<Command> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "AuthRepo")
+    }
+}
+
+impl<Command: AsyncCommands> AuthRepo<Command> {
+    fn new(redis_client: Command) -> Self {
+        Self { redis_client }
+    }
+
+    #[tracing::instrument]
+    async fn acquire_token_value(&mut self, token: String) -> Result<TokenValue, AuthError> {
+        // Acquire redis value from the key from the `token` parameter
+        let token_value: RedisResult<Option<String>> = self.redis_client.get(token).await;
+
+        match token_value {
+            Ok(Some(value)) => {
+                let token_value: TokenValue = serde_json::from_str(&value).unwrap();
+                Ok(token_value)
+            }
+            Ok(None) => Err(AuthError::TokenNotRegistered),
+            Err(redis_error) => {
+                sentry::capture_error(&redis_error);
+                Err(AuthError::FailedToAcquireToken(redis_error))
+            },
+        }
+    }
+
+    #[tracing::instrument]
+    async fn acquire_counter_limit(&mut self, token_value: &TokenValue) -> Result<i64, AuthError> {
+        let formatted_date = chrono::Utc::now().format("%Y-%m").to_string();
+        let key = format!("counter/{}/{}", formatted_date, token_value.user_email);
+        let value: RedisResult<Option<String>> = self.redis_client.get(key).await;
+
+        match value {
+            Ok(Some(value)) => {
+                let counter_limit: i64 = value.parse().unwrap();
```

Comment:
consider add error handling here. Add parse error in the enum and return something like 
`serde_json::from_str(&value).map_err(|e| AuthError::ParseError(e))`